src/core/response/parameters.rs

```diff
@@ -26,18 +28,26 @@ impl From<IdToken> for Json {
     }
 }
 
-// TODO: Update this type to something like:
-//
-// enum VpToken {
-//     Single(String),
-//     SingleAsMap(Map<String, Value>),
-//     Many(Vec<VpToken>),
-// }
-//
-// See: https://github.com/spruceid/oid4vp-rs/pull/8#discussion_r1750274969
-//
+/// OpenID Connect for Verifiable Presentations specification defines `vp_token` parameter:
+///
+/// > JSON String or JSON object that MUST contain a single Verifiable Presentation or
+/// > an array of JSON Strings and JSON objects each of them containing a Verifiable Presentations.
+/// >
+/// > Each Verifiable Presentation MUST be represented as a JSON string (that is a Base64url encoded value)
+/// > or a JSON object depending on a format as defined in Appendix A of [OpenID.VCI].
+/// >
+/// > If Appendix A of [OpenID.VCI] defines a rule for encoding the respective Credential
+/// > format in the Credential Response, this rules MUST also be followed when encoding Credentials of
+/// > this format in the vp_token response parameter. Otherwise, this specification does not require
+/// > any additional encoding when a Credential format is already represented as a JSON object or a JSON string.
+///
+/// See: [OpenID.VP#section-6.1-2.2](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-6.1-2.2)
 #[derive(Debug, Clone)]
-pub struct VpToken(pub String);
+pub enum VpToken {
+    Single(String),
```

Comment:
Based on the above comments, this should be a `Vec<u8>` instead of String, as the decoded value may not be UTF-8.

Reply:
> Each Verifiable Presentation MUST be represented as a JSON string (that is a Base64url encoded value)


Wouldn't this be encoded and not decoded? Is the inner `String` value not what is being sent over the wire?

Reply:
What is sent over the wire is `JSON::String(Base64Encoded(<Value>))`, for example the authorization response could be:

```
 HTTP/1.1 302 Found
  Location: https://client.example.org/cb#
    presentation_submission=...
    &vp_token=%22eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c%22
```

Reply:
Is the right way to think about this to include a method on the `VpToken` type itself for encoding the payload for the various variants. E.g.,

```rust
impl VpToken {
    pub fn as_base64_url_encoded(self) -> Result<String, Error> {
        let json: Json = self.try_into()?;
        let string = serde_json::to_string(&json)?;
        Ok(BASE64_URL_SAFE_NO_PAD.encode(string.as_bytes()))
    }
}
```

Reply:
> What is sent over the wire is `JSON::String(Base64UrlEncode(<Value>))`, for example the authorization response could be:
> 
> ```
>  HTTP/1.1 302 Found
>   Location: https://client.example.org/cb#
>     presentation_submission=...
>     &vp_token=%22ZXlKaGJHY2lPaUpJVXpJMU5pSXNJblI1Y0NJNklrcFhWQ0o5LmV5SnpkV0lpT2lJeE1qTTBOVFkzT0Rrd0lpd2libUZ0WlNJNklrcHZhRzRnUkc5bElpd2lhV0YwSWpveE5URTJNak01TURJeWZRLlNmbEt4d1JKU01lS0tGMlFUNGZ3cE1lSmYzNlBPazZ5SlZfYWRRc3N3NWM%22
> ```

Based on your comment above, is the following statement still accurate?

> Based on the above comments, this should be a Vec<u8> instead of String, as the decoded value may not be UTF-8.

Seems to me that `VpToken::Single` would contain `JSON::String(Base64UrlEncode(<Value>))`, which would be a `serde_json::Value` type and not a `Vec<u8>`.

Confused by your comments.